src/main/java/ch/njol/skript/lang/function/DefaultFunction.java

```diff
@@ -0,0 +1,311 @@
+
+		this.parameters = parameters;
+		this.execute = execute;
+		this.description = description;
+		this.since = since;
+		this.examples = examples;
+		this.keywords = keywords;
+	}
+
+	@Override
+	public T @Nullable [] execute(FunctionEvent<?> event, Object[][] params) {
+		Map<String, Object> args = new LinkedHashMap<>();
+
+		int length = Math.min(parameters.length, params.length);
+		for (int i = 0; i < length; i++) {
+			Object[] arg = params[i];
+			Parameter<?> parameter = parameters[i];
+
+			if (arg == null || arg.length == 0) {
+				if (parameter.isOptional()) {
+					continue;
+				} else {
+					return null;
+				}
+			}
+
+			if (arg.length == 1 || parameter.isSingleValue()) {
+				assert parameter.getType().getC().isAssignableFrom(arg[0].getClass())
+					: "argument type %s does not match parameter type %s".formatted(parameter.getType().getC().getSimpleName(),
+					arg[0].getClass().getSimpleName());
+
+				args.put(parameter.getName(), arg[0]);
+			} else {
+				assert parameter.getType().getC().isAssignableFrom(arg.getClass())
+					: "argument type %s does not match parameter type %s".formatted(parameter.getType().getC().getSimpleName(),
+					arg.getClass().getSimpleName());
+
+				args.put(parameter.getName(), arg);
+			}
+		}
+
+		FunctionArguments arguments = new FunctionArguments(args);
+		T result = execute.apply(arguments);
+
+		if (result == null) {
+			return null;
+		} else if (result.getClass().isArray()) {
+			//noinspection unchecked
+			return (T[]) result;
+		} else {
+			//noinspection unchecked
+			T[] array = (T[]) Array.newInstance(result.getClass(), 1);
+			array[0] = result;
+			return array;
+		}
+	}
+
+	@Override
+	public boolean resetReturnValue() {
+		return true;
+	}
+
+	/**
+	 * Returns this function's description.
+	 *
+	 * @return The description.
+	 */
+	public @NotNull String[] description() {
+		return description;
+	}
+
+	/**
+	 * Returns this function's version history.
+	 *
+	 * @return The version history.
+	 */
+	public @NotNull String[] since() {
+		return since;
+	}
+
+	/**
+	 * Returns this function's examples.
+	 *
+	 * @return The examples.
+	 */
+	public @NotNull String[] examples() {
+		return examples;
+	}
+
+	/**
+	 * Returns this function's keywords.
+	 *
+	 * @return The keywords.
+	 */
+	public @NotNull String[] keywords() {
+		return keywords;
+	}
+
+	public static class Builder<T> {
+
+		private final String name;
+		private final Class<T> returnType;
+		private final Map<String, Parameter<?>> parameters = new LinkedHashMap<>();
+
+		private Contract contract = null;
+
+		private String[] description, since, examples, keywords;
+
+		private Builder(@NotNull String name, @NotNull Class<T> returnType) {
+			Preconditions.checkNotNull(name, "name cannot be null");
+			Preconditions.checkNotNull(returnType, "return type cannot be null");
+
+			this.name = name;
+			this.returnType = returnType;
+		}
+
+		public Builder<T> contract(@NotNull Contract contract) {
```

Comment:
I would make use of Contract annotations, for example:
`@Contract(_ -> this)` - meaning, whatever values are passed, this method returns the object itself